pkg/rpcserver/rpcserver.go

```diff
@@ -232,17 +234,37 @@ func (serv *server) Serve(ctx context.Context) error {
 	g, ctx := errgroup.WithContext(ctx)
 	g.Go(func() error {
 		return serv.serv.Serve(ctx, func(ctx context.Context, conn *flatrpc.Conn) error {
-			return serv.handleConn(ctx, g, conn)
+			err := serv.handleConn(ctx, conn)
+			if err != nil {
+				log.Logf(0, "serv.handleConn returend %v", err)
+			}
+			return err
 		})
 	})
+	g.Go(func() error {
+		var info *handshakeResult
+		select {
+		case <-ctx.Done():
+			return nil
+		case info = <-serv.onHandshake:
+		}
+		serv.checkOnce.Do(func() {
```

Comment:
Do we need checkOnce now if we have a single goroutine that does runCheck once?

Reply:
Indeed. I've removed `checkOnce` and the nested goroutines.